src/Exchange.ts

```diff
@@ -0,0 +1,435 @@
+      return user.signMsg(orderHash)
+        .then(({ ecSignature }) => ({...order, ecSignature}))
+        .then(signedOrder => this.postRequest('exchange/order', signedOrder)
+          .then(() => ({
+            ...signedOrder,
+            hash: orderHash,
+            makerTokenAmount: utils.formatAmount(signedOrder.makerTokenAmount, makerDecimals),
+            takerTokenAmount: utils.formatAmount(signedOrder.takerTokenAmount, takerDecimals)
+          }))
+      )
+    } catch (error) {
+      return Promise.reject(error)
+    }
+  }
+
+  /**
+   * Prepares an on-chain transaction for a fill order.
+   * @param orderAddresses Array of order's maker, taker, makerToken, takerToken, and feeRecipient.
+   * @param orderValues Array of order's makerTokenAmount, takerTokenAmount, makerFee, takerFee, expirationTimestampInSec, and salt.
+   * @param fillTakerTokenAmount Desired amount of takerToken to fill.
+   * @param makerPath Path in the currency network of the maker token or [] if no currency network
+   * @param takerPath Path in the currency network of the taker token or [] if no currency network
+   * @param v ECDSA signature parameter v.
+   * @param r ECDSA signature parameter r.
+   * @param s ECDSA signature parameter s.
+   */
+  public async prepTakeOrder (
+    exchangeContractAddress: string,
+    makerAddress: string,
+    makerTokenAddress: string,
+    takerTokenAddress: string,
+    makerTokenValue: number | string,
+    takerTokenValue: number | string,
+    fillTakerTokenValue: number | string,
+    salt: string,
+    expirationUnixTimestampSec: string,
+    v: number,
+    r: string,
+    s: string,
+    {
+      gasLimit,
+      gasPrice,
+      makerTokenDecimals,
+      takerTokenDecimals
+    }: ExchangeOptions = {}
+  ): Promise<any> {
+    const { currencyNetwork, payment, transaction, user, utils } = this
+
+    try {
+      const [ makerDecimals, takerDecimals ] = await Promise.all([
+        currencyNetwork.getDecimals(makerTokenAddress, makerTokenDecimals),
+        currencyNetwork.getDecimals(takerTokenAddress, takerTokenDecimals)
+      ])
+      const feesRequest = {
+        exchangeContractAddress,
+        expirationUnixTimestampSec,
+        maker: makerAddress,
+        makerTokenAddress,
+        makerTokenAmount: utils.calcRaw(makerTokenValue, makerDecimals),
+        salt,
+        taker: ZERO_ADDRESS,
+        takerTokenAddress,
+        takerTokenAmount: utils.calcRaw(takerTokenValue, takerDecimals)
+      }
+      const { feeRecipient, makerFee, takerFee } = await this.getFees(feesRequest)
+      const makerPathObj = await currencyNetwork.isNetwork(makerTokenAddress)
+        ? await payment.getPath(
+          makerTokenAddress,
+          makerAddress,
+          user.address,
+          this.getPartialAmount(fillTakerTokenValue, takerTokenValue, makerTokenValue),
+          { decimals: makerTokenDecimals }
+        ) : {
+          path: [],
+          maxFees: 0,
+          estimatedGas: 21000,
```

Comment:
This is not enough gas for a tokenTransfer

Reply:
40000 should be a good number for a normal token transfer

Reply:
done